[v3] Improve empty document handling

felixfontein opened the pull request:
Parts of #684 that haven been fixed by @niemeyer in the meantime.

> Thanks, but we need to have all tests either in node_test.go or in the respective test files, using existing conventions. We have plenty of tests and test files already.

So far I found exactly one test for multi-document reading and writing, namely `TestEncoderMultipleDocuments` in `encode_test.go`. Where should I put the multi-document decoding and encoding tests with `Node`?

felixfontein committed:
Improve empty document handling.

felixfontein commented:
@niemeyer if you have time, could you at least take a quick look at the tests? Where should I put multi-document tests? Is a new file ok for that?

niemeyer commented:
@felixfontein Thanks for the changes. Node test seems nice. We still need to move those tests in yaml_test.go into their respective files. If you have a look around encode_test.go and decode_test.go you'll probably spot the patterns already in use.

felixfontein commented:
@niemeyer CI is failing with `Error: Unable to find Go version '1.16.0-beta1' for platform linux and architecture x64.`.

felixfontein commented:
Restarting CI.

niemeyer commented:
@felixfontein Sorry for being slow here. I'll reserve some time soon to do another push on go-yaml.

felixfontein commented:
@niemeyer I mainly wanted to re-run tests to see whether the semgrep issue is still around, but now it looks like you (or someone else) has to manually enable the workflow for this PR (GitHub changed policies for GHA to combat folks abusing GHA for crypto mining...).

felixfontein commented:
@niemeyer but thanks for trying to take another look at this :)

lollipopman commented:
@felixfontein & @niemeyer this pull request seems to work correctly against the current v3 branch, is it ready to be merged or is more work required? Would love to be able to preserve comments in empty yaml docs! Thanks!

felixfontein commented:
@lollipopman from my point of view, it is working correctly and can be merged. I'm not that familiar with the code base, though, so maybe I'm missing something. That's why it needs a review.

glopal commented:
@felixfontein @lollipopman Any update on this one? It looks like the build has some failures. Been waiting on this PR for a while now.

felixfontein commented:
@glopal IIRC the CI errors were unrelated to this PR. From my side this PR is still ready. It still needs review by the maintainer.

pivotaljohn commented:
In a delightful display of dominoes, merging and releasing this fix has the effect of making [SOPS](https://github.com/mozilla/sops) work seemlessly with [`ytt`](https://github.com/vmware-tanzu/carvel-ytt#readme) which is a growing combination among those plumbing sensitive data straight through processing pipelines (like folks doing automated deploys to Kubernetes). It [currently takes a hack](https://github.com/mozilla/sops/issues/936).

This is a subtly valuable fix for a (expanding) group I can point at.

Thanks for all your efforts, @felixfontein!! 🙏🏻

felixfontein commented:
@niemeyer I rebased since I saw that the broken workflow is gone.

pieterocp reviewed (commented):
~~One could probably un-comment the two tests in `stores/yaml/store_test.go`, unless they must only be un-commented post-merge?~~ (Ignore me, different projects, brain-fart)

pieterocp commented:
Any contributors willing to re-review this or does it need a rebase? Be nice to get this into 3.1.0 maybe

felixfontein commented:
It's already rebased to the current `v3` branch. The last change to the `v3` branch was two years ago.

ingydotnet committed:
Update the README.md for the YAML org fork

dims committed:
Sync with sigs.k8s.io/yaml/goyaml.v3

Signed-off-by: Davanum Srinivas <davanum@gmail.com>

dims committed:
Update GitHub Actions .github/workflows/go.yaml

Signed-off-by: Davanum Srinivas <davanum@gmail.com>

dims committed:
Consolidate patch(es) into existing files

Signed-off-by: Davanum Srinivas <davanum@gmail.com>

stefanprodan committed:
Fix go vet errors

Signed-off-by: Stefan Prodan <stefan.prodan@gmail.com>

stefanprodan committed:
Run go vet in CI

Signed-off-by: Stefan Prodan <stefan.prodan@gmail.com>

stefanprodan committed:
Format code with `go fmt`

Signed-off-by: Stefan Prodan <stefan.prodan@gmail.com>

stefanprodan committed:
Run `go fmt` in CI

Signed-off-by: Stefan Prodan <stefan.prodan@gmail.com>

scottrigby committed:
Change LICENSE and NOTICE to Apache standards

Most scanners including GitHub could not properly detect the previous format

- Adds full Apache License
- Combines notices from both files into existing NOTICE file
- The MIT licence from LICENSE is a permissively-licensed dependency
- Removed unneeded text. A pointer and a short summary notice is sufficient for
  MIT permissively-licensed dependency
- Follows Apache example NOTICE format per https://arc.net/l/quote/deyyhvsk

Signed-off-by: Scott Rigby <scott@r6by.com>

ingydotnet committed:
Update README to alert that the fork is pre-launch

sarahnovotny committed:
Merge pull request #13 from yaml/readme-status

Update README to alert that the fork is pre-launch

dims committed:
Switch to official vanity package name

Signed-off-by: Davanum Srinivas <davanum@gmail.com>

dims committed:
[v3] Formalize new package name

go.yaml.in/yaml/v3

ingydotnet committed:
Update the README.md for the launch release

Note: the import URLs and (possibly) the LICENSE details need to be
updated, but that will happen in separate commits.

zostay committed:
Add basic golangci-lint configuration

zostay committed:
Enable errcheck, ineffassign, and gofmt

Co-Authored-By: Ingy döt Net <ingy@ingy.net>
Co-Authored-By: ccoVeille <3875889+ccoVeille@users.noreply.github.com>

zostay committed:
Fix some gofmt issues

Removed redundant implied type mentions.

Co-Authored-By: Ingy döt Net <ingy@ingy.net>
Co-Authored-By: ccoVeille <3875889+ccoVeille@users.noreply.github.com>

zostay committed:
Fix ineffectual assignments

Setting a value to false in Go is redundant, so I just converted the
list of assignments of false to a bool declaration. The other case
involved removing an assignment to a variable that is never used again.

zostay committed:
Add .github/dependabot.yml

ingydotnet committed:
Add general purpose GNUmakefile

This Makefile ignores system Go and installs go locally in
`.git/.local/go-<version>/bin/go` and uses that for all make commands.

make <go-cmd>
make test
make test GO-VERSION=1.23.4

You can clone the repo and run `make test` without needing anything
installed (except make, curl and bash).

ingydotnet committed:
Add Makefile support for ys-test rule

ingydotnet committed:
Add CodeQL file suggested by GitHub

ingydotnet committed:
Rename .github/workflows/codeql.yml to use .yaml

ingydotnet committed:
Remove boilerplate comments from codeql.yaml

ingydotnet committed:
Reformat long lines in codeql.yaml

ingydotnet committed:
Remove unnecessary YAML quotes in codeql.yaml

ingydotnet committed:
Remove commented boilerplate from codeql.yaml

ingydotnet committed:
Update workflows/go.yaml for all branches

ingydotnet committed:
Use .cache/ for .makes/ and .local/ caching

ingydotnet committed:
Add `make` section to README.md

carloslima committed:
Handle decoding of `-0` from YAML

This is needed because Go constants cannot express -0.0:
https://staticcheck.dev/docs/checks/#SA4026

So when parsing `-0`/`-0.0` it would end up `0`/`0.0` in Go.

carloslima committed:
Add extra tests for roundtrip of "\t\n"

carloslima committed:
Handle roundtrip for `{"<<": []}`

carloslima committed:
Fix errors from alias/anchor with non alnum name

closes #44

carloslima committed:
bug: fix handling of '?' in unquoted values

closes #36

ingydotnet committed:
Update GHA to run on main, v4, v3, v2 and v1

carloslima committed:
Add tests against yaml-test-suite

We are testing against release `data-2022-01-17` of yaml-test-suite:
https://github.com/yaml/yaml-test-suite/tree/data-2022-01-17

carloslima committed:
Remove known failing tests after bugfix was merged

carloslima committed:
Run yaml-test-suite tests on PRs

carloslima committed:
Use standard testing library encode_test.go

carloslima committed:
Use standard testing library for decode_test.go

carloslima committed:
Use standard testing library for node_test.go

carloslima committed:
Use standard testing library for limit_test.go

carloslima committed:
Remove suite_test.go and gopkg.in/check.v1

carloslima committed:
Remove GO111MODULE from github workflow

Co-authored-by: ccoVeille <3875889+ccoVeille@users.noreply.github.com>

carloslima committed:
Use string literals to avoid escaping slashes

Co-authored-by: ccoVeille <3875889+ccoVeille@users.noreply.github.com>

andig committed:
Format all files with gofumpt

carloslima committed:
Change version to go.yaml.in/yaml/v4

We aren't going to release any non-security changes on v3, changes on
main will be part of v4

ingydotnet committed:
Finalize the GNUmakefile setup

ingydotnet committed:
Integrate YTS and GNUmakefile

carloslima committed:
Change test package name

ingydotnet committed:
Add a CONTRIBUTING.md file

carloslima committed:
Don't run yts tests using -race on github workflow

andig committed:
Fix unused types/breaks

carloslima committed:
YTS test instructs how to download test data

In order to run the yaml test suite we need the data files to be
present. A user with a fresh clone trying to run all tests will face an
error. This makes the error message clear and provides the step needed
to enable the tests to run.

andig committed:
Allow TypeError inspection and add ParserError

- allow non-terminal TypeError returning UnmarshalErrors exposing source location
- add terminal ParserError exposing source location

Co-authored-by: WGH <wgh@torlan.ru>

ingydotnet committed:
Tighten encoding style rules for literal scalars

Fixes the problem identified in:\
https://github.com/yaml/go-yaml/pull/46

ingydotnet committed:
Remove leading ws check for literal heuristics

carloslima committed:
Better reporting of results on yts tests

1. output a clear count of PASS and FAIL
2. exit with error when FAIL count differs from known
3. when a known failing test passes, display the test name and error out

ingydotnet committed:
Use unicode.IsSpace and add tests for it

carloslima committed:
Use helper functions to DRY up tests

This brings back the more expressive syntax on assertions while keeping
the implementation simple and not relying on external libraries

WGH- committed:
Fix TestUnmarshalerError test

The bug was both TestUnmarshalerError and TestObsoleteUnmarshalerError
were testing the same obsoleteFailingUnmarshaler, and
failingUnmarshaler remained untested. This likely slipped through during
rebase.

WGH- committed:
Handle TextUnmarshaler errors as type errors

Previously, errors returned by UnmarshalText were handled as hard
immediate errors, aborting unmarshaling immediately. Now they're treated
as TypeErrors, just like similar error conditions.

This is a follow-up to #19 which introduced TypeError. That PR had
yaml.Unmarshaler support contributed by me, but I forgot about
encoding.TextUnmarshaler. This commit amends that oversight.

ccoVeille committed:
Use Go 1.18 as minimum version

Use Go 1.18 as the minimum version for this project to take advantage of
the new features and improvements in the language.

We inherited Go 1.16 from old repository.

The last version of Go is 1.25, using Go 1.18 is reasonable.

ccoVeille committed:
Update CI with all supported versions

ccoVeille committed:
Fix repository URL in package comment

ccoVeille committed:
Fix Unwrap to support Go < 1.20

The code relies in the old way to unwrap things.
So it doesn't need an Unwrap method

This allows us to keep the feature without having to use Unwrap []error
and then need a //go:build tag.

Recent Go versions that support Unwrap []error still supports func
Is(error) bool and As(error) bool interfaces

Co-authored-by: WGH <wgh@torlan.ru>

inteon committed:
Renamed types to match Go naming conventions

Signed-off-by: Tim Ramlot <42113979+inteon@users.noreply.github.com>

ccoVeille committed:
Fix assert.ErrorIs arguments order

assert.ErrorIs should match errors.Is signature

testify assert.ErrorIs respects it

ccoVeille committed:
Simplify error message in assertion tests

Messages should be clear.
We must understand it's a regexp

ccoVeille committed:
Add ErrorAs assertion helper

assert.ErrorsAs will help when using errors.As.
It would bring clear messages, and will avoid panic when debugging.

ccoVeille committed:
Add missing tests for errors unwrapping

ccoVeille committed:
Improve the error unwrapping

TypeError can now unwrap errors wrapped in the UnmarshalError

ccoVeille committed:
Add a GitHub Action to check commit messages

The following rules were implemented:

The subject should not use the conventional commit format (e.g., feat, fix, etc.).
The subject should not start with square brackets.
The subject should start with a capital letter.
The subject should not end with a period.
The subject should not contain consecutive spaces.
The subject should not have trailing spaces.
The subject should be longer than 20 characters.
The subject should be shorter than 50 characters.
The subject and body should be separated by a single blank line.
Each line in the body should not have trailing spaces.

ccoVeille committed:
Add CI for linting YAML files

yamllint is used via a GitHub Actions workflow.

I configured its rules to comply with our project's YAML style guide.

ccoVeille committed:
Fix go.yaml linting issues against yamllint rules

ccoVeille committed:
Fix typos found in the codebase

ingydotnet committed:
Complete the LICENSE and NOTICE files

ingydotnet committed:
Use .yaml ext for dependabot config, reformat yaml

ingydotnet committed:
Remove v4 from workflow branches

The v4 branch no longer exists.

dependabot[bot] committed:
Bump actions/checkout from v4 to v5

Bumps the actions group with 1 update: [actions/checkout](https://github.com/actions/checkout).

Updates `actions/checkout` from 4 to 5
- [Release notes](https://github.com/actions/checkout/releases)
- [Changelog](https://github.com/actions/checkout/blob/main/CHANGELOG.md)
- [Commits](https://github.com/actions/checkout/compare/v4...v5)

---
updated-dependencies:
- dependency-name: actions/checkout
  dependency-version: '5'
  dependency-type: direct:production
  update-type: version-update:semver-major
  dependency-group: actions
...

Signed-off-by: dependabot[bot] <support@github.com>

ingydotnet committed:
SQUASH: Rename tooling/ to util/

I prefer util/ for these types of scripts.

Also in general I prefer short dir names that don't end with 's'
(plural).

ingydotnet committed:
SQUASH: Concerning double quotes

Double quotes are often over-used in Bash.
shellcheck won't complain if you do.
When in doubt, remove them and shee if shellcheck is still ok.
Only use them where needed.

Prefer single quotes to double quotes.
They have a different semantic.
When I see double quotes it's a signal that something is happening here.

In Bash, double quotes are critical in some places, and useless in
others.
Don't cargo cult. Learn the rules.

ingydotnet committed:
SQUASH: Be more strict

Always use /usr/bin/env to prefer a bash other than /bin/bash, which is
often too old.

Requiring Bash 4.4 should be no problem here.

People running locally on Mac can simply 'brew install bash'

ingydotnet committed:
SQUASH: Put all code into functions

Put main logic at top after strict and usage.

ingydotnet committed:
SQUASH: Introduce a die function

Use single char vars for color codes.

Replace `if ! foo; then bar; fi` with `foo || bar` in some places.

Better if usage doesn't exit. Can be used in more contexts.

ingydotnet committed:
SQUASH: Move helpers to util/common.bash

ingydotnet committed:
SQUASH: Replace $(cat ...) with $(< ...)

Surprised that shellcheck didn't call that one out.

ingydotnet committed:
SQUASH: Use true/false instead of 0/1

ingydotnet committed:
SQUASH: Remove unneeded flag variable

ingydotnet committed:
SQUASH: Use 'name() ( ... )' style functions

Using `(...)` instead of `{...}` defines function in a subprocess.

Reduces side effects. `local` is no longer needed.

ingydotnet committed:
SQUASH: Refactor error reporting code

ingydotnet committed:
SQUASH: Replace 'echo ... | foo' with 'foo <<<...'

ingydotnet committed:
SQUASH: Refactor some things to read cleaner

ingydotnet committed:
SQUASH: Simplify some conditions

`[[ $x ]]` is same as `[[ -n $x ]]` or `[[ $x != '' ]]`

ingydotnet committed:
SQUASH: Refactor error loop

ingydotnet committed:
SQUASH: Rename functions using _ with -

Function names can use any characters that command names can.

ingydotnet committed:
SQUASH: Rework error reporting for commits

This is simpler logic.

Made some style changes.

Tested a lot.

ingydotnet committed:
SQUASH: Make bash 4.4+ requirement explicit

garysferrao committed:
Change error message about decoding into Go map

YAML allows sequences and mappings as a key for a mapping.
Go does not support slices or maps as a map key; and the default
decoder returns an error "yaml: invalid map key %#v". but i feel
that this error message is misleading and may cause the user to
think that it's an error in the document.

i updated the error message to be indicate this situation and ask
the user to try `yaml.Node`

ccoVeille committed:
Add typos GitHub action

This will report typos in the codebase and documentation.

ccoVeille committed:
Add a linter for GitHub actions files

inteon committed:
Renamed functions to match Go naming conventions

Signed-off-by: Tim Ramlot <42113979+inteon@users.noreply.github.com>

inteon committed:
Apply review suggestions

Signed-off-by: Tim Ramlot <42113979+inteon@users.noreply.github.com>

inteon committed:
Apply review suggestions (2nd pass)

Signed-off-by: Tim Ramlot <42113979+inteon@users.noreply.github.com>

inteon committed:
Rename function names in comments too

Signed-off-by: Tim Ramlot <42113979+inteon@users.noreply.github.com>

inteon committed:
Use URI instead of Uri

Signed-off-by: Tim Ramlot <42113979+inteon@users.noreply.github.com>

ccoVeille committed:
Fix left behind typo

This way the CI passes.

ccoVeille committed:
Add gofumpt to our CI

gofumpt supersedes gofmt, there is no need to have them both.

ccoVeille committed:
Configure golangci-lint to report all errors

By default, golangci-lint reports only:
- 3 errors per linter.
- 50 errors in total.
- only one error per line (the first one encountered).

These settings are good default values when enabling golangci-lint to a
project with a lot of code and issues to fix.

But it comes with a drawback, when someone updates code, only a limited
number of errors will be reported, letting user think the code is OK.

I faced many times when people are pushing code, see errors reported by
the CI in GitHub Actions, and then fix them, but after pushing again,
new errors are reported, which were already present in the code before,
but were not reported because of the limits.

This creates frustrations, and people may feel that the linting process
is not effective.

Here we configure golangci-lint to report all errors by adjusting the
settings in the .golangci.yaml file.

ccoVeille committed:
Enable almost all govet rules in golangci-lint

govet is a powerful linter, unfortunately most rules are not enabled
by default.

The code quality in our project is good enough that we can enable almost
all govet rules without changes.

ccoVeille committed:
Enable dupwords in golangci-lint

This linter helps to catch repeated words in comments and strings.

Because of the example comments, we have some false-positives.
The golangci-lint configuration has been updated to include them.

The remaining errors reported by dupwords have been addressed.

ccoVeille committed:
Enable misspell in golangci-lint configuration

misspell helps to catch commonly misspelled words in the codebase.

Enabling the linter was straightforward. Only one typo was present.

ccoVeille committed:
Configure misspell linter to use American English

We are enforcing the use of American English spelling in our codebase to
maintain consistency and clarity.

The code and comments were updated to comply.

ccoVeille committed:
Enable nolintlint linter in golangci

This linter is about enforcing how the //nolint directive is used in the
codebase. This directive is used to disable some linter checks.

It's important to use this directive judiciously, and to enforce using
comments to explain why the directive is being used.

HadrienPatte committed:
Replace `reflect.Ptr` with `reflect.Pointer`

`Ptr` has been deprecated and has a [gofix](https://pkg.go.dev/golang.org/x/tools/gopls/internal/analysis/gofix#hdr-Constants) directive to inline it to `Pointer`.

See [`Ptr`](https://cs.opensource.google/go/go/+/refs/tags/go1.18:src/reflect/type.go;l=272).

Also replace `reflect.PtrTo` with `reflect.PointerTo` for the same
[reason](https://pkg.go.dev/reflect#PtrTo):

> PtrTo is the old spelling of PointerTo. The two functions behave identically.

ccoVeille committed:
Add file names linting to the CI

ls-lint is a powerful tool for linting file names.

It helps ensure that filenames follow a consistent naming convention.

Right now, we are only enforcing to disallow .yml files.
Only .yaml files are allowed.
